README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# colors3d

<!-- badges: start -->
<!-- badges: end -->

This R package provides functions to help generate two-dimensional and
three-dimensional color gradient legends.

The three main functions, `colors3d`, `colors2d`, and `colorwheel2d`
generate a color for each row of a user-supplied data set with 2-3
columns. These can then be used for plotting in various ways.

## Installation

You can install colors3d from GitHub with
`devtools::install_github("matthewkling/colors3d")` or from CRAN with
`install.packages("colors3d")`.

## Examples

Here’s a simple application of the three color mapping functions. This
example uses tidyverse, but this would all work in base R as well:

``` r
library(colors3d)
library(tidyverse)

# simulate a 3D data set
d <- expand_grid(x = 1:20, y = 1:20, z = 1:4)

# define and plot some 2D color mappings
d$colors2d <- colors2d(d[, 1:2])
d$colorwheel2d <- colorwheel2d(d[, 1:2])
d %>%
      gather(mapping, color, colors2d, colorwheel2d) %>%
      ggplot(aes(x, y, fill = color)) +
      facet_wrap(~mapping) +
      geom_raster() +
      scale_fill_identity()
```

<img src="man/figures/README-example-1.png" width="100%" />

``` r

# define and plot a 3D color mapping
d$color3d <- colors3d(d[, 1:3])
d %>%
      ggplot(aes(x, y, fill = color3d)) +
      facet_wrap(~z, nrow = 1, labeller = label_both) +
      geom_raster() +
      scale_fill_identity()
```

<img src="man/figures/README-example-2.png" width="100%" /> In a more
realistic application, we often want to create a pair of plots for a
given visualization: a “legend” in which the x and y dimensions match
those used to create the color mapping, and a second plot in which the
colors are then displayed in a different data space. This allows users
to understand relationships among four dimensions of the data (or 5, if
a 3D color mapping is used). Let’s use the `storms` dataset (from dplyr)
as an example, with hurricane windspeed, size, longitude, and latitude
as our variables of interest:

``` r
d <- na.omit(storms)
d$color <- colors2d(select(d, wind, hurricane_force_diameter),
                    xtrans = "rank", ytrans = "rank")

p1 <- ggplot(d, aes(wind, hurricane_force_diameter, color = color)) +
      geom_point() +
      scale_color_identity()
p2 <- ggplot() +
      geom_polygon(data = map_data("state"), 
                   aes(long, lat, group = group)) +
      geom_path(data = d, 
                aes(long, lat, color = color,
                    group = paste(name, year))) +
      scale_color_identity() +
      coord_cartesian(xlim = range(d$long),
                      ylim = range(d$lat))

library(patchwork)
p1 + p2
```

<img src="man/figures/README-example2-1.png" width="100%" />


NEWS.md
# colors3d 1.0.1.9000

* Added a `NEWS.md` file to track changes to the package.


R/colomap.R



#' Map values to a 3D legend in RGB colorspace.
#'
#' This function returns a color value for each row of the 3-column dataset
#' supplied, by transforming the input data and using it as RGB values.
#'
#' @param data Matrix or data frame with 3 numeric columns.
#' @param trans Either "fit" (defaut, histogram is rescaled) or "ecdf"
#'   (histogram is flattened).
#' @param order Integer from 1 to 6, each denoting a unique permutation of
#'   variables-to-color band mapping.
#' @param inversion Integer from 1 to 8, each denoting a unique combination of
#'   variables to invert.
#'   @param opacity Not currently used.
#' @return Vector of color values.

colors3d <- function(data, trans="fit", order=1, inversion=1, opacity=NULL){
      require(scales)
      require(combinat)
      data <- apply(data, 2, rescale)
      if(trans=="ecdf") data <- apply(data, 2, function(x)ecdf(x)(x))
      data <- data[,permn(1:3)[[order]]]
      invert <- (1:3)[as.logical(expand.grid(c(F,T), c(F,T), c(F,T))[inversion,])]
      data[,invert] <- 1- data[,invert]
      cols <- rep(NA, nrow(data))
      cols[which(is.finite(rowMeans(data)))] <- rgb(na.omit(data))
      cols
}



#' Map values to a 2D legend interpolated from 4 corner colors.
#'
#' This function returns a color value for each row of the 2-column dataset
#' supplied, based on a 2D color palette interpolated from 4 corner colors.
#'
#' @param data Matrix or data frame with 2 numeric columns; they will map to x
#'   and y.
#' @param colors Vector of 4 corner colors to interpolate, clockwise from upper
#'   right.
#' @return Vector of color values.

colors2d <- function(data, colors=c("yellow", "green", "blue", "magenta")){
      colors <- col2rgb(colors)/255
      data <- apply(data, 2, scales::rescale)
      interpolate <- function(i){
            x <- i[1]
            y <- i[2]
            x1 <- colors[,2] * x + colors[,3] * (1-x)
            x2 <- colors[,1] * x + colors[,4] * (1-x)
            x2 * y + x1 * (1-y)
      }
      rgb(t(apply(data, 1, interpolate)))
}





#' Internal function converting x-y do distance-angle.
#'
#' @param data Matrix or data frame with 2 numeric columns representing x and y.
#' @param xyratio Single number indicating unit ratio in x vs y direction.
#' @param xorigin, yorigin Numbers indicating center of polarization.
#' @return 2-column matrix of distances and angles.

polarize <- function(data, xyratio, xorigin=0, yorigin=0){
      distance <- sqrt((data[,1]-xorigin)^2 + ((data[,2]-yorigin) * xyratio)^2)
      angle <- acos((data[,1]-xorigin) / distance) * 180 / pi
      angle[data[,2]<yorigin] <- 360 - angle[data[,2]<yorigin]
      return(cbind(distance, angle))
}


#' Map values to a 2D colorwheel legend.
#'
#' This function returns a color value for each row of the 2-column dataset
#' supplied, based on a 2D color palette defined by a center color and a series
#' of peripheral colors.
#'
#' @param data Matrix or data frame with 2 numeric columns; they will map to x
#'   and y.
#' @param colors Vector of colors to interpolate: center followed by periphery
#'   counterclockwise from 3 o'clock.
#' @return Vector of color values.

colorwheel2d <- function(data, colors=c("black", "yellow", "green", "blue", "magenta"),
                         origin=NULL, xyratio=NULL){
      if(is.null(origin)) origin <- c(sum(range(data[,1], na.rm=T))/2,
                                      sum(range(data[,2], na.rm=T))/2)

      xrange <- range(data[,1], na.rm=T)
      yrange <- range(data[,2], na.rm=T)
      xmag <- plyr::round_any(max(abs(xrange)), (xrange[2]-xrange[1])/20, ceiling)
      ymag <- plyr::round_any(max(abs(yrange)), (yrange[2]-yrange[1])/20, ceiling)
      if(is.null(xyratio)) xyratio <- xmag / ymag

      pdata <- polarize(data, xyratio=xyratio, xorigin=origin[1], yorigin=origin[2])
      n <- length(colors)-1
      angle <- pdata[,2] / 360
      cl <- ceiling(angle * n) + 1
      fl <- floor(angle * n) + 1
      col <- matrix(NA, length(angle), 3)
      mx <- max(pdata[,1])

      colors <- col2rgb(colors)
      pal <- colors[,c(2:ncol(colors),2)] / 255
      center <- colors[,1] / 255
      center <- as.vector(center)

      for(i in 1:length(angle)){
            interp <- angle[i] * n - fl[i] + 1
            col_angle <- (as.vector(pal[,cl[i]]) * interp +
                                as.vector(pal[,fl[i]]) * (1-interp))
            col[i,] <- col_angle * pdata[,1][i] / mx + center * (1 - pdata[,1][i]/mx)
      }
      d <- rep(NA, nrow(data))
      d[which(!is.na(rowMeans(col)))] <- rgb(na.omit(col))
      return(d)
}



R/colormap.R



#' Map values to a 3D legend in RGB colorspace.
#'
#' This function returns a color value for each row of the 3-column dataset
#' supplied, by transforming the input data and using it as RGB values.
#'
#' @param data Matrix or data frame with 3 numeric columns.
#' @param trans Either "fit" (defaut, histogram is rescaled) or "ecdf"
#'   (histogram is flattened).
#' @param order Integer from 1 to 6, each denoting a unique permutation of
#'   variables-to-color band mapping.
#' @param inversion Integer from 1 to 8, each denoting a unique combination of
#'   variables to invert.
#'   @param opacity Not currently used.
#' @return Vector of color values.

colors3d <- function(data, trans="fit", order=1, inversion=1, opacity=NULL){
      require(scales)
      require(combinat)
      data <- apply(data, 2, rescale)
      if(trans=="ecdf") data <- apply(data, 2, function(x)ecdf(x)(x))
      data <- data[,permn(1:3)[[order]]]
      invert <- (1:3)[as.logical(expand.grid(c(F,T), c(F,T), c(F,T))[inversion,])]
      data[,invert] <- 1- data[,invert]
      cols <- rep(NA, nrow(data))
      cols[which(is.finite(rowMeans(data)))] <- rgb(na.omit(data))
      cols
}



#' Map values to a 2D legend interpolated from 4 corner colors.
#'
#' This function returns a color value for each row of the 2-column dataset
#' supplied, based on a 2D color palette interpolated from 4 corner colors.
#'
#' @param data Matrix or data frame with 2 numeric columns; they will map to x
#'   and y.
#' @param colors Vector of 4 corner colors to interpolate, clockwise from upper
#'   right.
#' @return Vector of color values.

colors2d <- function(data, colors=c("yellow", "green", "blue", "magenta"),
                     xtrans=c("none", "log", "rank"), ytrans=c("none", "log", "rank")){
      colors <- col2rgb(colors)/255

      if(xtrans=="rank") data[,1] <- ecdf(data[,1])(data[,1])
      if(ytrans=="rank") data[,2] <- ecdf(data[,2])(data[,2])
      if(xtrans=="log") data[,1] <- log(data[,1])
      if(ytrans=="log") data[,2] <- log(data[,2])

      data <- apply(data, 2, scales::rescale)
      interpolate <- function(i){
            x <- i[1]
            y <- i[2]
            x1 <- colors[,2] * x + colors[,3] * (1-x)
            x2 <- colors[,1] * x + colors[,4] * (1-x)
            x2 * y + x1 * (1-y)
      }
      rgb(t(apply(data, 1, interpolate)))
}





#' Internal function converting x-y do distance-angle.
#'
#' @param data Matrix or data frame with 2 numeric columns representing x and y.
#' @param xyratio Single number indicating unit ratio in x vs y direction.
#' @param xorigin, yorigin Numbers indicating center of polarization.
#' @return 2-column matrix of distances and angles.

polarize <- function(data, xyratio, xorigin=0, yorigin=0){
      distance <- sqrt((data[,1]-xorigin)^2 + ((data[,2]-yorigin) * xyratio)^2)
      angle <- acos((data[,1]-xorigin) / distance) * 180 / pi
      angle[data[,2]<yorigin] <- 360 - angle[data[,2]<yorigin]
      return(cbind(distance, angle))
}


#' Map values to a 2D colorwheel legend.
#'
#' This function returns a color value for each row of the 2-column dataset
#' supplied, based on a 2D color palette defined by a center color and a series
#' of peripheral colors.
#'
#' @param data Matrix or data frame with 2 numeric columns; they will map to x
#'   and y.
#' @param colors Vector of colors to interpolate: center followed by periphery
#'   counterclockwise from 3 o'clock.
#' @param origin Coorindates of color wheel center.
#' @param xyratio Scalar representing how to map the elliptical color wheel in
#'   the data space (the default 1 a circular mapping that weights the two
#'   dimensions equally).
#' @param kernel Optional function describing the shape of radial color
#'   gradients (default is a linear mapping corresponding to a triangular
#'   kernel); this function should take a vector of distances to the center as
#'   its sole input and return a positive number.
#' @return Vector of color values.

colorwheel2d <- function(data, colors=c("black", "yellow", "green", "cyan", "blue", "magenta", "red"),
                         origin=NULL, xyratio=NULL, kernel=NULL){
      result <- rep(NA, nrow(data))
      a <- which(!is.na(apply(data, 1, sum)))
      data <- na.omit(data)

      if(is.null(origin)) origin <- c(sum(range(data[,1], na.rm=T))/2,
                                      sum(range(data[,2], na.rm=T))/2)

      xrange <- range(data[,1])
      yrange <- range(data[,2])
      xmag <- plyr::round_any(max(abs(xrange)), (xrange[2]-xrange[1])/20, ceiling)
      ymag <- plyr::round_any(max(abs(yrange)), (yrange[2]-yrange[1])/20, ceiling)
      if(is.null(xyratio)) xyratio <- xmag / ymag

      pdata <- as.data.frame(polarize(data, xyratio=xyratio,
                                      xorigin=origin[1], yorigin=origin[2]))

      if(!is.null(kernel)) pdata$distance <- kernel(pdata$distance)
      pdata$angle <- pdata$angle / 360

      n <- length(colors)-1
      pdata$cl <- ceiling(pdata$angle * n) + 1
      pdata$fl <- floor(pdata$angle * n) + 1
      col <- matrix(NA, length(pdata$angle), 3)
      mx <- max(pdata$distance)

      colors <- col2rgb(colors)
      pal <- colors[,c(2:ncol(colors),2)] / 255
      center <- colors[,1] / 255
      center <- as.vector(center)

      getcol <- function(x){
            interp <- x[2] * n - x[4] + 1
            col_angle <- (as.vector(pal[,x[3]]) * interp +
                                as.vector(pal[,x[4]]) * (1-interp))
            col_angle * x[1] / mx + center * (1 - x[1]/mx)
      }

      col <- t(apply(pdata, 1, getcol))
      col[pdata$distance==0,] <- center
      result[a] <- rgb(col)
      return(result)
}


#' Palettes of dissimilar colors in RGB space.
#'
#' Many standard palette generators use only a slice of color space, which can
#' cause a lack of differentiability in palettes used to visualize categorical
#' factors with many levels. This function attempts to overcome this by
#' generating colors using nearest-neighbor distance maximization in 3D RGB
#' space.
#'
#' @param n Number of colors (integer).
#' @param res Number of distinct values in each RGB dimension (integer).
#' @param maxreps Max number of optimization iterations (integer).
#' @param radius Neighborhood size for potential moves, analagous to heating
#'   (integer).
distant_colors <- function(n, res=20, maxreps=100, radius=1){

      require(dplyr, quietly=T)
      require(FNN, quietly=T)

      f <- expand.grid(r=1:res,
                       g=1:res,
                       b=1:res)

      si <- sample_n(f, n, replace=F)

      for(i in 1:maxreps){

            si0 <- si

            for(j in 1:n){

                  # active location
                  sij<- si[j,]

                  # potential moves
                  hood <- filter(f,
                                 between(r, sij$r-radius, sij$r+radius),
                                 between(g, sij$g-radius, sij$g+radius),
                                 between(b, sij$b-radius, sij$b+radius))

                  # reference locations
                  sin <- si[-j,]

                  # find the move with max dist to nearest active location
                  dst <- get.knnx(sin, hood, k=1)$nn.dist
                  move <- hood[which.max(dst)[1],]

                  # execute optimal move
                  si[j,] <- move
            }

            # check for convergence
            if(all.equal(as.matrix(si0), as.matrix(si)) == T) break
      }

      if(i == maxreps) warning("Algorithm failed to converge, consider increasing maxreps parameter.")

      rgb(si, maxColorValue=res)
}


R/functions.R

#' Map values to a 3D legend in RGB colorspace.
#'
#' This function returns a color value for each row of the 3-column dataset
#' supplied, by transforming the input data and using it as RGB values.
#'
#' @param data Matrix or data frame with 3 numeric columns.
#' @param trans Either "none" (defaut, histogram is rescaled) or "rank"
#'   (histogram is flattened).
#' @param order Integer from 1 to 6, each denoting a unique permutation of
#'   variables-to-color band mapping. Under the default value of 1, the 
#'   three variables in `data` are respectively mapped onto the R, G, and B 
#'   bands of colorspace.
#' @param inversion Integer from 1 to 8, each denoting a unique combination of
#'   variables to reverse before mapping. Under the default value of 1, all 
#'   three variables are mapped with positive values at the high end of the
#'   color band. Together with the `order` parameter, this allows all possible
#'   48 unique mappings of a given set of variables onto 3D colorspace.
#' @param opacity Not currently used.
#' @returns Character vector of colors.
#' @examples 
#' d <- expand.grid(x = 1:49, y = 1:49)
#' d$z <- cos(sqrt((d$x-25)^2 + (d$y-25)^2))
#' plot(d[, 1:2], col = colors3d(d), pch = 15, cex = 2)
#' 
#' plot(d[, 1:2], col = colors3d(d, order = 2, inversion = 2), pch = 15, cex = 2)
#' @export
colors3d <- function(data, trans = "none", order = 1, inversion = 1, opacity = NULL){
      rescale <- function(x) (x - min(x)) / (max(x) - min(x))
      trans <- match.arg(trans, c("none", "rank"))
      data <- apply(data, 2, rescale)
      if(trans == "rank") data <- apply(data, 2, function(x) rank(x) / length(x))
      data <- data[, combinat::permn(1:3)[[order]]]
      invert <- (1:3)[as.logical(expand.grid(c(FALSE, TRUE), c(FALSE, TRUE), c(FALSE, TRUE))[inversion, ])]
      data[, invert] <- 1 - data[, invert]
      cols <- rep(NA, nrow(data))
      cols[which(is.finite(rowMeans(data)))] <- rgb(na.omit(data))
      cols
}



#' Map values to a 2D legend interpolated from 4 corner colors.
#'
#' This function returns a color value for each row of the 2-column dataset
#' supplied, based on a 2D color palette interpolated from 4 corner colors.
#'
#' @param data Matrix or data frame with 2 numeric columns; they will map to x
#'   and y.
#' @param colors Vector of 4 corner colors to interpolate, clockwise from upper
#'   right.
#' @param xtrans,ytrans Transformation to apply to x and y variables before
#'   applying a linear color mapping: either "none" (default), "log", or "rank".
#' @returns Character vector of colors.
#' @examples 
#' plot(iris, 
#'      col = colors2d(iris[, c("Sepal.Length", "Sepal.Width")]), 
#'      pch = 19, cex = 2)
#'      
#' plot(iris, 
#'      col = colors2d(iris[, c("Sepal.Length", "Sepal.Width")],
#'                     colors = c("limegreen", "gold", "black", "dodgerblue"),
#'                     xtrans = "rank", ytrans = "rank"), 
#'      pch = 19, cex = 2)
#' @export
colors2d <- function(data, 
                     colors = c("yellow", "green", "blue", "magenta"),
                     xtrans = c("none", "log", "rank"), 
                     ytrans = c("none", "log", "rank")){
      
      xtrans <- match.arg(xtrans, c("none", "log", "rank"))
      ytrans <- match.arg(ytrans, c("none", "log", "rank"))
      
      colors <- col2rgb(colors)/255

      if(xtrans == "rank") data[, 1] <- rank(data[, 1]) / nrow(data)
      if(ytrans == "rank") data[, 2] <- rank(data[, 2]) / nrow(data)
      if(xtrans == "log") data[, 1] <- log(data[, 1])
      if(ytrans == "log") data[, 2] <- log(data[, 2])

      data <- apply(data, 2, scales::rescale)
      interpolate <- function(i){
            x <- i[1]
            y <- i[2]
            x1 <- colors[, 2] * x + colors[, 3] * (1-x)
            x2 <- colors[, 1] * x + colors[, 4] * (1-x)
            x2 * y + x1 * (1-y)
      }
      
      cols <- rep(NA, nrow(data))
      cols[which(is.finite(rowMeans(data)))] <- rgb(t(apply(na.omit(data), 1, interpolate)))
      cols
}



#' Internal function converting x-y do distance-angle.
#'
#' @param data Matrix or data frame with 2 numeric columns representing x and y.
#' @param xyratio Single number indicating unit ratio in x vs y direction.
#' @param xorigin,yorigin Numbers indicating center of polarization.
#' @returns 2-column matrix of distances and angles.
polarize <- function(data, xyratio, xorigin=0, yorigin=0){
      distance <- sqrt((data[,1]-xorigin)^2 + ((data[,2]-yorigin) * xyratio)^2)
      angle <- acos((data[,1]-xorigin) / distance) * 180 / pi
      angle[data[,2]<yorigin] <- 360 - angle[data[,2]<yorigin]
      return(cbind(distance, angle))
}


#' Map values to a 2D colorwheel legend.
#'
#' This function returns a color value for each row of the 2-column dataset
#' supplied, based on a 2D color palette defined by a center color and a series
#' of peripheral colors.
#'
#' @param data Matrix or data frame with 2 numeric columns; they will map to x
#'   and y.
#' @param colors Vector of colors to interpolate: center followed by periphery
#'   counterclockwise from 3 o'clock.
#' @param origin Coordindates of color wheel center.
#' @param xyratio Scalar representing how to map the elliptical color wheel in
#'   the data space (the default 1 a circular mapping that weights the two
#'   dimensions equally).
#' @param kernel Optional function describing the shape of radial color
#'   gradients (default is a linear mapping corresponding to a triangular
#'   kernel); this function should take a vector of distances to the center as
#'   its sole input and return a positive number.
#' @returns Character vector of colors.
#' @examples
#' plot(iris, 
#'       col = colorwheel2d(iris[, c("Sepal.Length", "Sepal.Width")]), 
#'       pch = 19, cex = 2)
#'       
#' plot(iris, 
#'       col = colorwheel2d(
#'                   iris[, c("Sepal.Length", "Sepal.Width")],
#'                   origin = c(5.5, 2.5),
#'                   kernel = function(x) x ^ .5), 
#'       pch = 19, cex = 2)
#' @export
colorwheel2d <- function(data, 
                         colors = c("black", "yellow", "green", "cyan", "blue", "magenta", "red"),
                         origin = NULL, xyratio = NULL, kernel = NULL){
      result <- rep(NA, nrow(data))
      a <- which(!is.na(apply(data, 1, sum)))
      data <- na.omit(data)

      if(is.null(origin)) origin <- c(sum(range(data[,1], na.rm=T))/2,
                                      sum(range(data[,2], na.rm=T))/2)

      xrange <- range(data[,1])
      yrange <- range(data[,2])
      xmag <- plyr::round_any(max(abs(xrange)), (xrange[2]-xrange[1])/20, ceiling)
      ymag <- plyr::round_any(max(abs(yrange)), (yrange[2]-yrange[1])/20, ceiling)
      if(is.null(xyratio)) xyratio <- xmag / ymag

      pdata <- as.data.frame(polarize(data, xyratio=xyratio,
                                      xorigin=origin[1], yorigin=origin[2]))
      names(pdata) <- c("distance", "angle")

      if(!is.null(kernel)) pdata$distance <- kernel(pdata$distance)
      pdata$angle <- pdata$angle / 360

      n <- length(colors)-1
      pdata$cl <- ceiling(pdata$angle * n) + 1
      pdata$fl <- floor(pdata$angle * n) + 1
      col <- matrix(NA, length(pdata$angle), 3)
      mx <- max(pdata$distance)

      colors <- col2rgb(colors)
      pal <- colors[,c(2:ncol(colors),2)] / 255
      center <- colors[,1] / 255
      center <- as.vector(center)

      getcol <- function(x){
            interp <- x[2] * n - x[4] + 1
            col_angle <- (as.vector(pal[,x[3]]) * interp +
                                as.vector(pal[,x[4]]) * (1-interp))
            col_angle * x[1] / mx + center * (1 - x[1]/mx)
      }

      col <- t(apply(pdata, 1, getcol))
      col[pdata$distance==0,] <- center
      result[a] <- rgb(col)
      return(result)
}


#' Palettes of dissimilar colors in RGB space.
#'
#' Many standard palette generators use only a slice of color space, which can
#' cause a lack of differentiability in palettes used to visualize categorical
#' factors with many levels. This function attempts to overcome this by
#' generating colors using nearest-neighbor distance maximization in 3D RGB
#' space.
#'
#' @param n Number of colors (integer).
#' @param res Number of distinct values in each RGB dimension (integer).
#' @param maxreps Max number of optimization iterations (integer).
#' @param radius Neighborhood size for potential moves, analagous to heating.
#' @param avoid_white Logical, default is TRUE.
#' @param seed Integer used to seed randomization during search; leave as NULL to generate different results each time, or set a value to generate reproducible results.
#' @returns Character vector of colors.
#' @examples
#' plot(runif(20), runif(20), 
#'      col = distant_colors(20), 
#'      pch = 16, cex = 3)
#' @export
distant_colors <- function(n, res=20, maxreps=1000, radius=10, avoid_white=TRUE, seed=NULL){

      if(!is.null(seed)) set.seed(seed)
      
      if(avoid_white) n <- n + 1

      f <- expand.grid(r=1:res,
                       g=1:res,
                       b=1:res)

      si <- f[sample(nrow(f), n),]

      for(i in 1:maxreps){

            si0 <- si

            for(j in 1:n){

                  # active location
                  sij<- si[j,]

                  # potential moves
                  h <- f[f$r > sij$r-radius & f$r < sij$r+radius &
                                  f$g > sij$g-radius & f$g < sij$g+radius &
                                  f$b > sij$b-radius & f$b < sij$b+radius,]

                  # reference locations
                  sin <- si[-j,]

                  # find the move with max dist to nearest active location
                  dst <- FNN::get.knnx(sin, h, k=1)$nn.dist
                  move <- h[which.max(dst)[1],]

                  # execute optimal move
                  si[j,] <- move
            }

            # check for convergence
            if(all.equal(as.matrix(si0), as.matrix(si)) == TRUE) break
      }

      if(i == maxreps) warning("Algorithm failed to converge, consider increasing maxreps parameter.")

      if(avoid_white) si <- si[setdiff(1:nrow(si), which.max(si$r + si$g + si$b)),]

      rgb(si, maxColorValue=res)
}
